pkg/server/file.go

```diff
@@ -123,15 +123,44 @@ func (t *TranslateServiceServer) UploadTranslationFile(
 		return nil, status.Errorf(codes.Internal, "")
 	}
 
-	// If the uploaded file is original and populateTranslations flag is true, populate the translated messages.
-	if messages.Original && params.populateTranslations {
-		allMessages, err := t.repo.LoadMessages(ctx, params.serviceID, repo.LoadMessagesOpts{})
+	// When updating original messages, changes might affect translations - transform and update all translations.
+	if messages.Original {
+		all, err := t.repo.LoadMessages(ctx, params.serviceID, repo.LoadMessagesOpts{})
 		if err != nil {
 			return nil, status.Errorf(codes.Internal, "")
 		}
 
-		if err = t.populateTranslatedMessages(ctx, params.serviceID, messages, allMessages); err != nil {
-			return nil, err
+		all = all.Replace(*messages)
+		prev, _ := all.SplitOriginal()
+
+		// Find original messages with altered text, then replace text in associated messages for all translations.
+		all = t.alterTranslations(all, findUntranslatedIDs(*prev, *messages))
+
+		// If populateMessages is true - populate missing messages for all translations.
+		if params.populateTranslations {
+			all = t.populateTranslations(all)
+		}
+
+		// Fuzzy translate untranslated messages for all translations
+		if all, err = t.fuzzyTranslate(ctx, all); err != nil {
+			return nil, fmt.Errorf("fuzzy translate messages :%w", err)
+		}
+
+		// Update all translations
+		for i := range all {
+			if all[i].Original {
+				continue
+			}
+
+			err = t.repo.SaveMessages(ctx, params.serviceID, &all[i])
+			switch {
+			default:
+				// noop
+			case errors.Is(err, repo.ErrNotFound):
+				return nil, status.Errorf(codes.NotFound, "service not found")
+			case err != nil:
+				return nil, status.Errorf(codes.Internal, "")
+			}
```

Comment:
I am almost sure that it is possible to move this from `if messages.Original` block, with minor changes on outside block, thus reducing repetition